pages/api/applications/[applicationId]/stages/index.ts

```diff
@@ -1,5 +1,133 @@
+import { Nullable } from '../../../../../types/utils';
 
-export default function handler(req: NextApiRequest, res: NextApiResponse) {
-  res.status(200).end(`Request sent to ${req.url}\n`);
+const { ERROR, SUCCESS } = StatusMessageType;
+const prisma = new PrismaClient();
+
+type ApplicationDataWithStages = Application & { applicationStages: ApplicationStage[] };
+
+enum MessageType {
+  INVALID_TYPE,
+  INVALID_DATE,
+  INVALID_APPLICATION_ID,
+  APPLICATION_NOT_FOUND,
+  CREATED,
+}
+
+const messages = Object.freeze({
+  [MessageType.INVALID_TYPE]: {
+    type: ERROR,
+    message:
+      'Application Stage type is invalid. Type must be one of: ' + Object.values(ApplicationStageType).join(', ') + '.',
+  },
+  [MessageType.INVALID_DATE]: {
+    type: ERROR,
+    message: 'Application Stage date is invalid.',
+  },
+  [MessageType.INVALID_APPLICATION_ID]: {
+    type: ERROR,
+    message: 'Query parameter applicationId (.../applications/:applicationId/...) is invalid.',
+  },
+  [MessageType.APPLICATION_NOT_FOUND]: {
+    type: ERROR,
+    message: 'User does not have any application with the given ID.',
+  },
+  [MessageType.CREATED]: { type: SUCCESS, message: `Application stage has been created.` },
+});
+
+async function handler(
+  uid: string,
+  req: NextApiRequest,
+  res: NextApiResponse<ApiResponse<ApplicationStageData | ApplicationDataWithStages>>,
+) {
+  const errorMessageIfInvalid = validatePathParameters(req);
+  if (errorMessageIfInvalid !== null) {
+    return res.status(HttpStatus.BAD_REQUEST).json(createJsonResponse({}, messages[errorMessageIfInvalid]));
+  }
+
+  const application = await getApplication(uid, Number(req.query.applicationId));
+  if (application === null) {
+    return res.status(HttpStatus.NOT_FOUND).json(createJsonResponse({}, messages[MessageType.APPLICATION_NOT_FOUND]));
+  }
+
+  switch (req.method) {
+    case HttpMethod.POST:
+      return postHandler(application, req, res);
+    case HttpMethod.GET:
+      return getHandler(application, res);
+    default:
+      return rejectHttpMethod(res, req.method);
+  }
+}
+
+async function postHandler(
+  application: ApplicationDataWithStages,
+  req: NextApiRequest,
+  res: NextApiResponse<ApiResponse<ApplicationStageData>>,
+) {
+  const paramValidationError = validatePostRequest(req);
+  if (paramValidationError !== null) {
+    return res.status(HttpStatus.BAD_REQUEST).json(createJsonResponse({}, messages[paramValidationError]));
+  }
+
+  const { type, date } = req.body;
+  const parameters: ApplicationStagePostData = { type, date: new Date(date) };
+
+  const newStage: ApplicationStageData = await prisma.applicationStage.create({
+    data: { ...parameters, applicationId: application.id },
+    select: { applicationId: true, id: true, type: true, date: true, emojiUnicodeHex: true, remark: true },
+  });
+  return res.status(HttpStatus.CREATED).json(createJsonResponse(newStage, messages[MessageType.CREATED]));
 }
+
+async function getHandler(
```

Comment:
Why is the GET /api/applications/:id/stages returning the application?

Should this be done in GET /api/applications/:id instead?

Reply:
We also have no need for this route.

Reply:
Ah.. yes 😞